python_graphql_client/graphql_client.py

```diff
@@ -13,10 +13,11 @@
 class GraphqlClient:
     """Class which represents the interface to make graphQL requests through."""
 
-    def __init__(self, endpoint: str, headers: dict = None):
+    def __init__(self, endpoint: str, headers: dict = {}, options: dict = {}):
```

Comment:
Do you have a preference for `options` over `**kwargs` here? I think storing it on the instance as `options` still makes sense.

So this function would turn into:

```python
def __init__(self, endpoint: str, headers: dict = {}, **kwargs: int):
  ...
  self.options = kwargs
```

And the execute function changes to:

```python
def execute(
  self,
  query: str,	       
  variables: dict = None,	        
  operation_name: str = None,	      
  headers: dict = None,
  **kwargs: int,
):
  ...
  result = requests.post(
    ...
    **{self.options, **kwargs}
  )
```
```

Reply:
@steve148 I considered the following three approaches.

- `verify: bool, cert: Union[str, Sequence[str]], auth: ...`
- `options: dict = {}`
- `**kwargs: int`

I decided on the second option because the first option wasn't scalable and the third option, I would say, felt too flexible/generous and potentially, will get in our way when we want to add other parameters in the future.
I found the second option most balanced and explicit but I don't feel strongly about it.
Please let me know if you see any upsides in using `**kwargs` in this case.

Reply:
Option number 1 is definitely too rigid, we definitely don't want to be maintaining an exact replica of the arguments allowed by the different libraries we use under the hood.

I personally feel like `**kwargs` is a better choice of `options` since `**kwargs` is the more commonly used pattern for creating flexible functions in python. I would rather use the more common pattern over one that we come up with just for this project.

From my view, `execute` is just a nice wrapper overtop of `requests.post`. The distinguishing arguments to `execute` are `query`, `variables`, and `operation_name`. Looking at it now, `headers` isn't related to graphql but instead comes from the `requests` library. This will cause a breaking change in this function's interface, but `headers` should probably be made part of the `**kwargs` in the future after this change lands.

Either way, I think going forward we should split arguments into two groups:

* Arguments like `query` which are related to graphql and the functionality this project provides. These arguments should be defined individually in each function (aka `def execute(query: str, ...):`
* Arguments like `verify` which are related to the underlying communication protocol and aren't specific to graphql or this library. These arguments should be defined as a group in `**kwargs` so the interface is flexible to changes to the underlying library.

Reply:
@steve148 That makes sense. I updated the API to use `**kwargs` as suggested. I'm not sure how we can also take `headers` through `**kwargs` though. Would it be easy to determine which argument should be treated as `headers` and which ones should not when everything is passed together?

Reply:
🙏  thanks @DaleSeo! I'm probaably thinking a few too many steps ahead when I mention `headers` being part of `**kwargs`. One core piece of functionality we do have right now is being able to set headers on the class instance which gets used by all calls. This was pretty useful to us and probably isn't something we want to get rid of.

My thinking would be you could still set `headers` the class instantation and in the method call.

```py
client = GraphqlClient(..., headers={ 'shared_key': 'shared_value' })

client.execute(..., headers={'another_key': 'another_value'})
```

And then the HTTP call would be a merge of the two dictionaries

```py
requests.post(..., headers={'shared_key': 'shared_value', 'another_key': 'another_value'})
```

I think something like the following would achieve this idea https://stackoverflow.com/questions/20656135/python-deep-merge-dictionary-data. At the same time, this is starting to get complex and may not align with the goal of this project to be a simple graphql client.